src/analyses/base.ml

```diff
@@ -2569,6 +2576,52 @@ struct
     (* D.join ctx.local @@ *)
     ctx.local
 
+  let unassume (ctx: (D.t, _, _, _) ctx) e =
+    let e_cpa = CPA.bot () in
+    let vars = Basetype.CilExp.get_vars e |> List.unique ~eq:CilType.Varinfo.equal in
+    if List.for_all (fun v -> not v.vglob) vars then (
+      let e_cpa = List.fold_left (fun e_cpa v ->
+          CPA.add v (VD.top_value v.vtype) e_cpa
+        ) e_cpa vars
+      in
+      (* TODO: structural unassume instead of invariant hack *)
+      let e_d =
+        (* The usual recursion trick for ctx. *)
+        (* Must change ctx used by ask to also use new st (not ctx.local), otherwise recursive EvalInt queries use outdated state. *)
+        (* Note: query is just called on base, but not any other analyses. Potentially imprecise, but seems to be sufficient for now. *)
+        let local: D.t = {ctx.local with cpa = e_cpa} in
+        let rec ctx' asked =
+          { ctx with
+            ask = (fun (type a) (q: a Queries.t) -> query' asked q)
+          ; local
+          }
+        and query': type a. Queries.Set.t -> a Queries.t -> a Queries.result = fun asked q ->
+          let anyq = Queries.Any q in
+          if Queries.Set.mem anyq asked then
+            Queries.Result.top q (* query cycle *)
+          else (
+            let asked' = Queries.Set.add anyq asked in
+            match q with
+            | MayEscape _ -> false (* HACK: avoids get_var in inv_exp thinking non-first var is global *)
```

Comment:
There's a major problem here in base for using `invariant` on a top state to construct the unassumed state. Since `invariant` is not executed on `ctx.local`, but requires a `ctx`, a new `ctx` has to be constructed with the altered state. Doing so also means that queries inside `invariant` are only answered by base itself, so multithreadedness and escape queries always return top, causing evaluations inside `invariant` to treat locals as globals (and call privatization), which breaks everything.

Reply:
To not be unsound, I made this delegate some queries outside to `ctx.ask`, but others are answered by base using the modified local state.